# vmrseq: Detecting variably methylated regions (VMRs) from single-cell bisulfite sequencing

The R package `vmrseq` is a novel computational tool developed for pinpointing variably methylated regions (VMRs) in scBS-seq data without prior knowledge on size or location. High-throughput single-cell measurements of DNA methylation allows studying inter-cellular epigenetic heterogeneity, but this task faces the challenges of sparsity and noise. vmrseq overcomes these challenges and identifies variably methylated regions accurately and robustly. 

![](figures/method.png)


## Installation

You can install the development version of `vmrseq` from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("nshen7/vmrseq")
```

## License/Copyright
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) 
This package is made available under an MIT license.  
