---
title: "Analyzing single-cell bisulfite sequencing data with vmrseq"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analyzing single-cell bisulfite sequencing data with vmrseq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
devtools::load_all()
```

# Installation

This package will be submitted to Bioconductor. For now, you can install the development version from our [GitHub repository](https://github.com/nshen7/vmrseq) through running this:

```{r, eval = FALSE}
# install.packages("devtools")
devtools::install_github("nshen7/vmrseq")
```

After installation, load the package:

```{r setup, eval = FALSE}
library(vmrseq)
```

You might also have to manually install one of the dependencies of vmrseq, the `bumphunter` package (using `BiocManager::install('bumphunter')`). 

# Apply vmrseq to single-cell BS-seq data

```{r}
data(se0)
GenomicRanges::granges(se0)
```

On a side note: Individual-cell files can be processed into SummarizedExperiment object (with NA-dropped sparseMatrix representation) using `data.pool` function, see:
```{r eval=FALSE}
?data.pool
```


```{r}
gr <- vmrseq.smooth(se0)
output <- vmrseq.fit(gr)
```

