\nonstopmode{}
\documentclass[letterpaper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `vmrseq'}}
\par\bigskip{\large \today}
\end{center}
\inputencoding{utf8}
\ifthenelse{\boolean{Rd@use@hyper}}{\hypersetup{pdftitle = {vmrseq: Probabilistic Modeling of Single-cell Methylation Heterogeneity}}}{}
\begin{description}
\raggedright{}
\item[Type]\AsIs{Package}
\item[Title]\AsIs{Probabilistic Modeling of Single-cell Methylation Heterogeneity}
\item[Version]\AsIs{0.99.0}
\item[Author]\AsIs{Ning Shen}
\item[Maintainer]\AsIs{Ning Shen }\email{ning.shen@stat.ubc.ca}\AsIs{}
\item[Description]\AsIs{High-throughput single-cell measurements of DNA methylation allows studying inter-cellular epigenetic heterogeneity, but this task faces the challenges of sparsity and noise. We present vmrseq, a statistical method that overcomes these challenges and identifies variably methylated regions accurately and robustly.}
\item[License]\AsIs{MIT + file LICENSE}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[RoxygenNote]\AsIs{7.2.3}
\item[biocViews]\AsIs{}
\item[Depends]\AsIs{R (>= 3.5.0)}
\item[Imports]\AsIs{bumphunter, dplyr, BiocParallel, DelayedArray, GenomicRanges,
ggplot2, methods, tidyr, locfit, gamlss.dist, recommenderlab,
HDF5Array, data.table, SummarizedExperiment, IRanges, S4Vectors}
\item[Suggests]\AsIs{knitr, rmarkdown, testthat (>= 3.0.0)}
\item[Config/testthat/edition]\AsIs{3}
\item[VignetteBuilder]\AsIs{knitr}
\item[URL]\AsIs{}\url{https://nshen7.github.io/vmrseq/}\AsIs{}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{cell\_1}{cell\_1}{cell.Rul.1}
\keyword{datasets}{cell\_1}
%
\begin{Description}
This dataset is an example of a single-cell file that can be input to
\code{vmrseq::data.pool} function. It contains the first 100 CpG sites in
a cell from mouse frontal cortext dataset published by Luo et al. (2017).
\end{Description}
%
\begin{Usage}
\begin{verbatim}
data(cell_1)
\end{verbatim}
\end{Usage}
%
\begin{Format}
A data frame with 100 rows and 5 variables (no column names):
\begin{description}

\item[V1] Chromosome
\item[V2] Genomic coordinate
\item[V3] Strand information
\item[V4] Number of methylated reads
\item[V5] Number of reads in total

\end{description}

\end{Format}
%
\begin{References}
Luo, Chongyuan et al. \emph{Single-cell methylomes identify neuronal
subtypes and regulatory elements in mammalian cortex.}. Science (New York, N.Y.)
vol. 357,6351 (2017): 600-604.
\end{References}
%
\begin{Examples}
\begin{ExampleCode}
data(cell_1)
cell_1

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{computeVar}{Compute the smoothed variance}{computeVar}
%
\begin{Description}
Compute the smoothed variance
\end{Description}
%
\begin{Usage}
\begin{verbatim}
computeVar(gr, M, meanMeth, bpWindow, sparseNAdrop, parallel)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gr}] same as in \code{vmrseq.smooth}

\item[\code{M}] numeric matrix of binary single-cell methylation status. Row number
should be equal to the number of CpG sites and column number should be equal
to the number of cells.

\item[\code{meanMeth}] numeric vector of across-cell mean methylation. Length should
be equal to the number of CpG sites.

\item[\code{bpWindow}] same as in \code{vmrseq.smooth}

\item[\code{sparseNAdrop}] same as in \code{vmrseq.smooth}

\item[\code{parallel}] logical value that indicates whether function should be
run in parallel
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{computeVarCutoff}{Compute cutoff on smoothed variance for determining candidate regions}{computeVarCutoff}
%
\begin{Description}
Compute cutoff on smoothed variance for determining candidate regions
\end{Description}
%
\begin{Usage}
\begin{verbatim}
computeVarCutoff(alpha, meth, total, pars_u, pars_m, n = 1e+05)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{alpha}] level of significance

\item[\code{meth}] vector of meth read counts

\item[\code{total}] vector of total read counts

\item[\code{pars\_u}] parameters used in the ZIBB distribution for unmethylated grouping

\item[\code{pars\_m}] parameters used in the BB distribution for methylated grouping

\item[\code{n}] number of simulations
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{data.pool}{Pool single-cell file together into an HDF5-based SummarizedExperiment object with sparse matrix representation.}{data.pool}
%
\begin{Description}
This function pools individual-cell CpG read files into a
SummarizedExperiment object so that it can be input to the \code{vmrseq.smooth}
function. Note that in each cell, sites with hemimethylation or intermediate
methylation levels (i.e., 0 < meth\_read/total\_read < 1) will be removed.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
data.pool(
  cellFiles,
  sep,
  writeDir,
  chrNames,
  colData = NULL,
  sparseNAdrop = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cellFiles}] Vector of character strings indicating single-cell file
paths you wish to pool into SE object(s). Cell files should be in BED-like
format, where the first 5 columns in each file must be: <chr>, <pos>, <strand>,
<meth\_read>, <total\_read>, in strict order. Each row contains info of one CpG.
Strand is assumed to be properly filpped. Cell files can be zipped ones in .gz
format.

\item[\code{sep}] the field separator character. Values on each line of cell file are
separated by this character.

\item[\code{writeDir}] A single character string indicating a folder directory where
you wish to store the processed SE object(s). The SE will be stored in HDF5 format.

\item[\code{chrNames}] Single or a vector of character strings representing chromosome
names in cell files. Only chromosomes listed in \code{selectChrs} will be
processed.

\item[\code{colData}] A DataFrame or data.frame object containing colData for the SE
object (applied to all chromosomes).

\item[\code{sparseNAdrop}] Logical value indicating whether or not to use NA-dropped
sparseMatrix representation. 'NA-dropped' means replacing NAs as zeros and
then represent 0 as a very small positive value close to 0 so that the data
matrix is converted to a traditional sparseMatrix form with 0 as default
entry value. (see \code{sparseNAMatrix-class} from package \code{recommenderlab}).
Using NA-dropped sparseMatrix representation helps to save RAM during data
processing, as well as storage space saving to disk. Currently only supports
TRUE!!
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Directly write out to the `writeDir` and does not return anything.
\end{Value}
\inputencoding{utf8}
\HeaderA{extractCoord}{Extract genomic coordinates of a particular chromosome in a cell file.}{extractCoord}
%
\begin{Description}
Extract genomic coordinates of a particular chromosome in a cell file.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
extractCoord(file, chr, sep)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{file}] file path

\item[\code{chr}] chromosome name

\item[\code{sep}] separator between columns
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{extractInfo}{Extract and process methylation info of a particular chromosome from a cell file.}{extractInfo}
%
\begin{Description}
Extract and process methylation info of a particular chromosome from a cell file.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
extractInfo(file, chr, sep)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{file}] file path

\item[\code{chr}] chromosome name

\item[\code{sep}] separator between columns
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{fillNA}{Fill NA's in missing cites per cell}{fillNA}
%
\begin{Description}
Fill NA's in missing cites per cell
\end{Description}
%
\begin{Usage}
\begin{verbatim}
fillNA(file, chr, pos_full, sep)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{file}] file path

\item[\code{chr}] chromosome name

\item[\code{pos\_full}] a complete list of genomic coordinates

\item[\code{sep}] separator between columns
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{HDF5NAdrop2matrix}{Coerce a NA-dropped HDF5 matrix back to regular matrix form where NAs are not dropped to 0.}{HDF5NAdrop2matrix}
%
\begin{Description}
Coerce a NA-dropped HDF5 matrix back to regular matrix form where NAs are not
dropped to 0.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
HDF5NAdrop2matrix(hdf5_assay)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{hdf5\_assay}] an DelayedMatrix object with dropped NA values (e.g. an
assay from HDF5SummarizedExpriment object saved by vmrseq::data.pool)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Returns a matrix.
\end{Value}
\inputencoding{utf8}
\HeaderA{region.summary}{Compute regional average methylation for individual cells.}{region.summary}
%
\begin{Description}
Compute regional average methylation for individual cells.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
region.summary(SE, region_ranges, sparseNAdrop = is_sparse(assays(SE)[[1]]))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{SE}] \code{SummarizedExperiment} object with one (and only one) assay that
contains *binary* methylation status of CpG sites in individual cells. In usual
analysis workflow (of vmrseq), \code{SE} should be the output of \code{vmrseq::data.pool}.

\item[\code{region\_ranges}] \code{GRanges} object that contains genomic coordinates
of regions of interest.

\item[\code{sparseNAdrop}] logical value that represents whether the NA values are
droppped in the input \code{SE} object. \code{SE} objects output by
\code{vmrseq::data.pool} are NA dropped. See \code{?vmrseq::data.pool}
for details about NA-dropped representation.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Returns a \code{SummarizedExperiment} object that contains the regional
average methylation per cell.
\end{Value}
\inputencoding{utf8}
\HeaderA{smoothMF}{Kernel smoothing function}{smoothMF}
%
\begin{Description}
Kernel smoothing function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
smoothMF(x, y, weights, chr, minInSpan, bpSpan, verbose, parallel)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] vector of the x values

\item[\code{y}] vector of the y values

\item[\code{weights}] vector of the weight associated with each data point

\item[\code{chr}] string of the chromosome name

\item[\code{minInSpan}] minimum number of sites in the span

\item[\code{bpSpan}] base pair of the span

\item[\code{verbose}] logical value that indicates whether progress messages
should be printed to stdout

\item[\code{parallel}] logical value that indicates whether function should be
run in parallel
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{tp.estimate}{Estimate transition probilities conditioning on CpG-CpG distance}{tp.estimate}
%
\begin{Description}
This function first computes transition probilities
conditioning on CpG-CpG distance in each cell. Then the probs from
individual cells are smoothed over CpG-CpG distance using `loess`
with inverse-variance fitting.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
tp.estimate(
  list,
  max_dist_bp = 2000,
  buffer_bp = 3000,
  lags = 1:10,
  BPPARAM = bpparam(),
  degree = 2,
  span = 0.02,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{list}] a list of data.frame objects. Each data.frame
contains information of 1 unit of training data (can be a cell or a
subtype) and should have 3 columns in strict rder of: (chr), (pos),
(binary methyl value). Column names are not necessary.

\item[\code{max\_dist\_bp}] positive integer value indicating the maximum
CpG-CpG distance in base pairs before the transition probabilities
reach constant value. Default value is 2000.

\item[\code{buffer\_bp}] length of buffer in base pairs used to fit smoothing
curve near maximum distance and plot diagnostics. Default is 3000.

\item[\code{lags}] a vector indicating possible number of lagged CpGs for
estimation of transition probabilities.

\item[\code{BPPARAM}] a \code{BiocParallelParam} object to specify the parallel
backend. The default option is \code{BiocParallel::bpparam()} which will
automatically creates a cluster appropriate for the operating system.

\item[\code{degree}] 'degree' argument for `loess` function

\item[\code{span}] 'span' argument for `loess` function

\item[\code{...}] additional arguments passed into the `loess` function.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a 'transitProbs' object. Postfixes rule in the output variables:
P(0|0) => '00'; P(0|1) => '01'; P(1|0) => '10'; P(1|1) => '11'.
\end{Value}
\inputencoding{utf8}
\HeaderA{tp.plot}{Plot transition probability distribution}{tp.plot}
%
\begin{Description}
Plot transition probability distribution
\end{Description}
%
\begin{Usage}
\begin{verbatim}
tp.plot(tp, line_size = 0.2, plot_train = T, point_size = 0.2)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{tp}] 'transitProbs' object storing information about trained transition
probabilities. Can be obtained from function `estimTransitProbs`

\item[\code{line\_size}] size of fitted loess smooth line. Default value is 0.2.

\item[\code{plot\_train}] logical value indiating whether to plot training data.
Default is TRUE.

\item[\code{point\_size}] size of training data points. Only applicable when
`plot\_train = T`. Default value is 0.2.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A plot of the transition probability distribution
\end{Value}
\inputencoding{utf8}
\HeaderA{vmrseq.fit}{Construct candidate regions and detect variably methylated regions.}{vmrseq.fit}
%
\begin{Description}
Construct candidate regions (CRs) by taking groups of consecutive loci that exceed
threshold on the variance of smoothed relative methylation levels and detect
variably methylated regions (VMRs) by optimizing a hidden Markov model (HMM).
\end{Description}
%
\begin{Usage}
\begin{verbatim}
vmrseq.fit(
  gr,
  alpha = 0.05,
  maxGap = 2000,
  stage1only = FALSE,
  minNumCR = 5,
  minNumVMR = 5,
  gradient = TRUE,
  tp = NULL,
  control = vmrseq.optim.control(),
  verbose = TRUE,
  BPPARAM = bpparam()
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gr}] \code{GRanges} object output by \code{vmrseq::vmrseq.smooth},
containing genomic coordinates (chr, start, end) and summarized information
(meth, total, var) of CpG sites in the input dataset.

\item[\code{alpha}] positive scalar value between 0 and 1 that represents the
designated significance level for determining variance threshold of candidate
regions construction. The variance threshold is determined by taking 1-alpha
quantile value of an approximate null distribution of variance simulated
from the beta priors of emission probability in the hidden Markov model.
Default value of alpha is 0.05.

\item[\code{maxGap}] integer value representing maximum number of base pairs in
between neighboring CpGs to be included in the same VMR. Default value is
2000 bp.

\item[\code{stage1only}] boolean value indicating whether the algorithm should run
stage 1 of vmrseq (the construction of candidate regions) only. If set to TRUE,
the function will output only the candidate regions. Default is FALSE.

\item[\code{minNumCR}] positive integer value representing the minimum number of
CpG sites within a candidate region. Default value is 5.

\item[\code{minNumVMR}] positive integer value representing the minimum number of
CpG sites within a variably methylated region. Default value is 5.

\item[\code{gradient}] logical value indicating whether exponentiated gradient
descent shall be applied to update prevalence parameter. Default is TRUE. If
set as FALSE, initial values (i.e., value of \code{inits} arguments in
\code{vmrsqe::vmrseq.optim.control}, can be set up in the \code{control}
argument of this function) are used as prevalence parameter for decoding
hidden states.

\item[\code{tp}] a `transitProbs-class` object that contains the transition
probability distribution used for HMM optimization. Default value is
transition probability \code{vmrseq:::tp0} built in the package that was
previously trained on mouse brain cells. See manuscript for training
procedure and data source.

\item[\code{control}] list of miscellaneous parameters used to control optimization
of the HMM model. Default is output of \code{vmrseq::vmrseq.optim.control()}.
Can be changed by tweaking arguments in function \code{vmrseq::vmrseq.optim.control()}.

\item[\code{verbose}] logical value that indicates whether progress messages
should be printed to stdout. Defaults value is TRUE.

\item[\code{BPPARAM}] a \code{BiocParallelParam} object to specify the parallel
backend. The default option is \code{BiocParallel::bpparam()} which will
automatically creates a cluster appropriate for the operating system.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
The results object is a list of 6 elements that contains the following information:
1. `gr`: The `Granges` object that has been input to `vmrseq.fit` with two added metadata columns:
+ `cr\_index` = Index in reference to rows of `cr.ranges`, denoting row number of the candidate region to which the CpG site belongs.
+ `vmr\_index` = Index in reference to rows of `vmr.ranges`, denoting row number of the variably methylated region to which the CpG site belongs.
2. `vmr.ranges`: A `Granges` object with the coordinates of each detected variably methylated region (each row is a VMR), with metadata columns:
+ `num\_cpg` = Number of observed CpG sites in the VMR.
+ `start\_ind` = Index of the starting CpG sites in reference to rows of `gr`.
+ `end\_ind` = Index of the ending CpG sites in reference to rows of `gr`.
+ `pi` = Prevalence of the methylated grouping (see manuscript for details)
+ `loglik\_diff` = Difference in log-likelihood of two-grouping and one-grouping HMM fitted to the VMR; can be used to rank the VMRs.
3. `cr.ranges`: A `Granges` object with the coordinates of each candidate region (each row is a candidate region), with metadata column:
+ `num\_cpg` = Number of observed CpG sites in the candidate region.
4. `alpha`: Designated significance level (default 0.05, can be changed by user with function argument). It is used for determining the threshold on variance used for constructing candidate. The threshold is computed by taking the (1-alpha) quantile of an approximate null distribution of variance (see manuscript for details).
5. `var\_cutoff`: Variance cutoff computed from `alpha`.
6. `bb\_params`: Beta-binomial parameter used in emission probability of the HMM model; they are determined by the magnitude of the input dataset (see manuscript for details).
\end{Value}
\inputencoding{utf8}
\HeaderA{vmrseq.optim.control}{Auxiliary function as user interface for vmrseq optimization.}{vmrseq.optim.control}
%
\begin{Description}
Typically only used when calling vmrseq function with the
option \code{control}.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
vmrseq.optim.control(
  inits = c(0.2, 0.5, 0.8),
  epsilon = 0.001,
  backtrack = T,
  eta = ifelse(backtrack, 0.05, 0.005),
  maxIter = 100
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{inits}] vector of numeric values between 0 and 1 representing initial
values of pi\_1 shall be taken in optimization algorithm.

\item[\code{epsilon}] numeric value representing the convergence upper bound for
the algorithm.

\item[\code{backtrack}] logical value indicating whether to use backtracking line
search to automatically adjust learning rate. Default is TRUE.

\item[\code{eta}] a numeric value representing the learning rate in optimization.
Default is \code{ifelse(backtrack, 0.05, 0.005)}.

\item[\code{maxIter}] positive integer value representing the maximum number of
iterations in optimization algorithm.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
the list of arguments for optimization control
\end{Value}
\inputencoding{utf8}
\HeaderA{vmrseq.smooth}{Smoothing on single-cell bisulfite sequencing data for the purpose of constructing candidate regions.}{vmrseq.smooth}
%
\begin{Description}
\code{vmrseq.smooth} takes a \code{SummarizedExperiment} object
with information of methylation level of individual cells as input, and
perform a kernel smoother to ‘relative’ methylation levels of individual
cells prior to constructing candidate regions. Purpose of the smoothing is
to adjust for uneven coverage biases and borrow information from nearby sites.
See manuscript for detailed description.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
vmrseq.smooth(
  SE,
  bpWindow = 2000,
  sparseNAdrop = is_sparse(assays(SE)[[1]]),
  verbose = TRUE,
  BPPARAM = bpparam()
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{SE}] \code{SummarizedExperiment} object with one (and only one) assay that
contains *binary* methylation status of CpG sites in individual cells. We
recommend using output by \code{vmrseq::data.pool} (i.e., an NA-dropped
HDF5-based SummarizedExperiment object) to prevent running out of memory.

\item[\code{bpWindow}] positive integer that represents the width (in bp) of
smoothing window. Default value is 2000.

\item[\code{sparseNAdrop}] logical value that represents whether the NA values are
droppped in the input \code{SE} object. \code{SE} objects output by
\code{vmrseq::data.pool} are NA dropped. See \code{?vmrseq::data.pool}
for details about NA-dropped representation.

\item[\code{verbose}] logical value that indicates whether progress messages
should be printed to stdout. Defaults value is TRUE.

\item[\code{BPPARAM}] a \code{BiocParallelParam} object to specify the parallel
backend. The default option is \code{BiocParallel::bpparam()} which will
automatically creates a cluster appropriate for the operating system.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a \code{GRanges} object that contains the result of smoothing.
The object retains genomic coordinates (chr, start, end) of input CpG
sites, in the same order as in the input \code{SE} object. Three
column are added (on top of original metadata columns for the CpG sites in
\code{SE}, if any):
1. meth: methylated cell count of the CpG
2. total: total (non-missing) cell count of the CpG
3. var: variance computed based on individual-cell smoothed relative methylation levels.
\end{Value}
%
\begin{SeeAlso}
\code{\LinkA{data.pool}{data.pool}}, \code{\LinkA{vmrseq.fit}{vmrseq.fit}}
\end{SeeAlso}
\printindex{}
\end{document}
