#' @title Main function for detecting and evaluating significance of VMRs in
#' single-cell bisulfite sequencing data.
#'
#' @description Performs a two-step approach that (1) detects candidate regions,
#' and (2) detect variably methylated region(s) inside each candidate region
#' using a hidden Markov model.
#'
#' @param SE SummarizedExperiment object with one and *only* one assay that
#' contains *binary* methylation status of CpG sites in individual cells. We
#' recommend using HDF5-based SummarizedExperiment object to prevent running
#' out of memory.
#' @param cutoff positive scalar value that represents the cutoff value of
#' variance that is used to discover candidate regions. Default value is 0.10.
#' @param penalty
#' @param maxGap integer value representing maximum number of basepairs in
#' between neighboring CpGs to be included in the same VMR.
#' @param bpSpan a positive integer that represents the length in base pairs of
#' the smoothing span window if \code{smooth} is TRUE. Default value is 1000.
#' @param minInSpan positive integer that represents the minimum number of CpGs
#' in a smoothing span window if \code{smooth} is TRUE. Default value is 10.
#' @param tp transitProbs object that contains estimated transition
#' probabilities. Can be obtained by the 'estimTransitProbs' function. If
#' \code{tp==NULL}, internal transition probabilities in \code{vmrseq} is used.
#' Default is NULL.
#' @param maxNumMerge positive integer that represents the maximum number of
#' CpGs between two VMRs that can be tolerated when merging VMRs in the same
#' candidate region. Default is 0.
#' @param minNumLong positive integer that represents the minimum number of
#' CpGs to consider for a *long* candidate region. For fast computation use.
#' Default is 20. Long regions will be performed a more thorough search of
#' optimized prevalence value. Minimum value is the \code{minNumRegion}.
#' @param gradient should gradient descent be applied to optimize pi? If
#' set to FALSE, \code{vmrseq.control()$inits} will be used as candidate values
#' for pi, but no gradient descent will be performed.
#' @param control this sets the control parameters of the outer iterations
#' algorithm. The default setting is the \code{vmrseq.control} function.
#' @param verbose logical value that indicates whether progress messages should
#' be printed to stdout. Defaults value is TRUE.
#' @param BPPARAM a \code{BiocParallelParam} object to specify the parallel
#' backend. The default option is \code{BiocParallel::bpparam()} which will
#' automatically creates a cluster appropriate for the operating system.
#' @param minNumCR
#' @param minNumVMR
#' @param bpWindow
#'
#' @return a list of two \code{GRanges} object that contains the results of the
#' inference.
#'
#' @importFrom BiocParallel bplapply register MulticoreParam bpparam
#' @importFrom bumphunter clusterMaker getSegments
#' @importFrom stats fitted median
#' @importFrom gamlss.dist dZIBB dBB
#' @importFrom locfit locfit lp
#' @importFrom DelayedArray rowSums colSums rowMeans colMeans
#' @import dplyr
#' @import GenomicRanges
#'
#'
#' @export
#'
#' @examples
#'
#'
vmrseq <- function(SE,
                   cutoff = 0.1, # param for CR calling
                   penalty = 0, # params for VMR calling
                   maxGap = 1000, minNumCR = 5, minNumVMR = 5, # params for VMR calling
                   bpWindow = 2000, # param for individual-cell smoother
                   bpSpan = 1000, minInSpan = 10, # params for across-cell smoother
                   tp = NULL,
                   maxNumMerge = 0, minNumLong = 0,
                   gradient = TRUE,
                   control = vmrseq.control(),
                   verbose = TRUE, BPPARAM = bpparam()) {

  if (is.null(cutoff) | length(cutoff) != 1 | cutoff <= 0)
    stop("'cutoff' has to be a postive scalar value.")
  # if (length(penalty)!=1 | penalty < 0)
  #   stop("'penalty' has to be a non-negative scalar value.")
  # if (minNumRegion < 3)
  #   stop("'minNumRegion' must be at least 3.")
  # if (minNumLong < minNumRegion)
  #   stop("'minNumLong' must be greater or equal to `minNumRegion`.")
  if (!is.logical(gradient))
    stop("'gradient' must be a logical value (TRUE or FALSE).")
  if (!class(SE)[1] %in% c("RangedSummarizedExperiment", "SummarizedExperiment"))
    stop("'SE' must be a SummarizedExperiment object.")
  if (!is.null(tp) & class(tp) != "transitProbs")
    stop("'tp' must be a transitProbs object. Can be estimated by function `estimTransitProbs`.")

  # # Check that GRanges object is sorted
  # if (is.unsorted(gr)) {
  #   message("'gr' is not sorted. Sorting 'gr' now.")
  #   gr <- sort(gr)
  # }

  for (chromosome in unique(seqnames(SE))) {
    SE_chr <- subset(SE, seqnames(SE) == chromosome)
    if (min(diff(start(SE_chr))) < 2)
      stop("There exists at least 2 rows with position difference less than 2 bp.")
  }

  values(SE)$meth <- rowSums(assays(SE)[[1]], na.rm = T)
  values(SE)$total <- rowSums(assays(SE)[[1]]>=0, na.rm = T)

  if (min(values(SE)$total) < 3)
    message("We suggest removing CpG sites with across-cell coverage lower than 3.")

  # Register the parallel backend
  BiocParallel::register(BPPARAM)
  backend <- paste0("BiocParallel:", class(bpparam())[1])

  if (bpparam()$workers == 1) {
    if (verbose) {
      mes <- "Parallel: Using a single core (backend: %s)."
      message(sprintf(mes, backend))
    }
    parallel <- FALSE
  } else {
    if (verbose) {
      mes <- paste0("Parallel: Parallelizing using %s workers/cores ",
                    "(backend: %s).")
      message(sprintf(mes, bpparam()$workers, backend))
    }
    parallel <- TRUE
  }

  # Bumphunt candidate regions
  message(
    "Step 1: Detecting candidate regions with (smoothed) variance larger than ",
    cutoff, "..."
  )
  # Outputs list of index vectors. Each list element is one CR.
  res_cr <- callCandidRegion(
    SE = SE,
    cutoff = cutoff,
    maxGap = maxGap, minNumCR = minNumCR,
    bpWindow = bpWindow,
    minInSpan = minInSpan, bpSpan = bpSpan,
    maxNumMerge = maxNumMerge,
    verbose = verbose,
    parallel = parallel
  )

  # Indexes of candidate regions
  CRI <- res_cr$CRI

  cr_index <- rep(NA, length(SE))
  cr_index[unlist(CRI)] <- rep.int(1:length(CRI), lengths(CRI))

  # Add summary stats (smoothed var and CR index) into output
  values(SE) <- cbind(values(SE), res_cr$mf_smooth, data.frame(var = res_cr$var), cr_index)

  # Percentage of sites in CRs
  pct_incr <- round(sum(lengths(CRI))/length(SE)*100, 2)
  if (is.null(CRI)) {
    message("...No candidate regions pass the cutoff of ", unique(abs(cutoff)))
    return(NULL)
  } else {
    message("...Finished calling candidate regions - found ", length(CRI),
            " candidate regions in total.
  ...", pct_incr,
            "% QC-passed sites are called to be in candidate regions.")
  }

  if (pct_incr <= 10) {
    message("WARNING: Consider lowering 'cutoff' since only ", pct_incr,
            "% QC-passed sites are called to be in candidate region.
    ...If not, might induce low statistical power.")
    warning("Consider lowering 'cutoff' since only ", pct_incr,
            "% QC-passed sites are called to be in candidate region.")
  }

  message(
    "Step 2: Detecting VMRs",
    ifelse(penalty > 0, yes = " with", no = " without"),
    " penalty",
    ifelse(penalty > 0, yes = paste0("=", round(penalty, 2)), no = ""),
    "..."
  )
  t1 <- proc.time()

  # Outputs a GRanges objects with VMR ranges and summary information
  VMRI <- searchVMR(
    gr = granges(SE),
    CRI = CRI,
    penalty = penalty,
    maxGap = maxGap, minNumVMR = minNumVMR,
    tp = tp,
    maxNumMerge = maxNumMerge,
    minNumLong = minNumLong,
    gradient = gradient,
    control = control,
    verbose = verbose,
    parallel = parallel
  )

  t2 <- proc.time()

  if (length(VMRI) == 0) {
    message("No VMR detected.")
    return(NULL)
  } else {
    message("...Finished detecting VMRs - took ",
            round((t2 - t1)[3]/60, 2), " min and ",
            nrow(VMRI), " VMRs found in total.
  ...", round(sum(VMRI$end_ind-VMRI$start_ind+1) / length(SE) * 100, 2),
            "% QC-passed sites are called to be in VMRs.")
  }

  vmr.gr <- indexToGranges(gr = granges(SE), index = VMRI, type = "VMR")
  cr.gr <- indexToGranges(gr = granges(SE), index = CRI, type = "CR")

  VMRI2 <- lapply(1:nrow(VMRI), function(i) VMRI$start_ind[i]:VMRI$end_ind[i]) # list of indices
  vmr_index <- rep(NA, length(SE))
  vmr_index[unlist(VMRI2)] <- rep.int(1:length(VMRI2), lengths(VMRI2))
  values(SE) <- cbind(values(SE), vmr_index, values(vmr.gr)[vmr_index,])

  return(list(gr_qced = granges(SE), VMRs = vmr.gr, CRs = cr.gr))
}



